plugins/plugin-java/che-plugin-java-server/src/main/java/org/eclipse/che/plugin/java/languageserver/JavaLanguageServerLauncher.java

```diff
@@ -83,18 +87,30 @@ public JavaLanguageServerLauncher(
       NotifyJsonRpcTransmitter notifyTransmitter,
       EventService eventService,
       ProjectManager projectManager,
-      ProjectsSynchronizer projectSynchronizer) {
+      ProjectsSynchronizer projectSynchronizer,
+      JavaLanguageServerExtensionService javaService) {
     this.processorJsonRpcCommunication = processorJsonRpcCommunication;
     this.executeCliendCommandTransmitter = executeCliendCommandTransmitter;
     this.notifyTransmitter = notifyTransmitter;
     this.eventService = eventService;
     this.projectManager = projectManager;
     this.projectSynchronizer = projectSynchronizer;
+    this.javaService = javaService;
+    this.notificationHandler = Executors.newSingleThreadExecutor();
 
     launchScript = Paths.get(System.getenv("HOME"), "che/ls-java/launch.sh");
   }
 
+  @PreDestroy
+  public void shutDown() {
+    notificationHandler.shutdown();
```

Comment:
I don't really care about what happens after shutdown is initiated. No reason to wait for it, really.

Reply:
It is not about whether you really care about something or not, it is more about generally accepted norms of code writing and patterns that are used throughout Eclipse Che project.

Reply:
Sorry, but "someone else is doing it this way", is not an argument. Can you give a factual reason why I should wait for the shutdown? I _am_ interested.

Reply:
It is not "someone else" - it is "the Eclipse Che community". However if you don't see the functional benefit and disrespect community coding norms I will try to give a better explanation.

You're saying that there is "no reason to wait for it, really", but my remark was not about waiting of tasks termination, it was about task termination itself.

According to `ExecutorService` javadoc `shutdown()` does: 
>Initiates an orderly shutdown in which previously submitted tasks are executed, but no new tasks will be accepted.

In other words you code will not terminate tasks being executed or submitted. Since your thread is not a `daemon` and you run external methods like `projectManager.update(registeredProject)`, the termination of tasks being executed may be delayed or task execution may be infinite depending on those methods implementations which in turn may lead to unexpected consequences. Thus this is a possible place of failure and the idea of requested changes is exactly to avoid introduction of bogus source code.

The solution that is suggested is not "someone's else" but an option that is recommended by an Oracle API documentation: 
```java
 void shutdownAndAwaitTermination(ExecutorService pool) {
   pool.shutdown(); // Disable new tasks from being submitted
   try {
     // Wait a while for existing tasks to terminate
     if (!pool.awaitTermination(60, TimeUnit.SECONDS)) {
       pool.shutdownNow(); // Cancel currently executing tasks
       // Wait a while for tasks to respond to being cancelled
       if (!pool.awaitTermination(60, TimeUnit.SECONDS))
           System.err.println("Pool did not terminate");
     }
   } catch (InterruptedException ie) {
     // (Re-)Cancel if current thread also interrupted
     pool.shutdownNow();
     // Preserve interrupt status
     Thread.currentThread().interrupt();
   }
 }
```
You can read more here: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html